File: python/10_maximum_depth_of_binary_tree.py
"""
Q10. Maximum Depth of Binary Tree (Category: Trees)

Problem: Find the maximum depth (height) of a binary tree.

Approach: Recursive DFS — depth = 1 + max(left depth, right depth).

Time Complexity : O(n)
Space Complexity: O(h) where h is tree height
"""
from __future__ import annotations
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0,
                 left: Optional["TreeNode"] = None,
                 right: Optional["TreeNode"] = None):
        self.val = val
        self.left = left
        self.right = right


def max_depth(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))


if __name__ == "__main__":
    #     3
    #    / \
    #   9  20
    #     /  \
    #    15    7
    root = TreeNode(3, TreeNode(9),
                    TreeNode(20, TreeNode(15), TreeNode(7)))
    print(max_depth(root))  # 3

File: cpp/10_MaximumDepthOfBinaryTree.cpp
/*
 * Q10. Maximum Depth of Binary Tree (Category: Trees)
 *
 * Problem: Find the maximum depth (height) of a binary tree.
 *
 * Approach: Recursive DFS — depth = 1 + max(left, right).
 *
 * Time Complexity : O(n)
 * Space Complexity: O(h) where h is tree height
 */
#include <iostream>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

int maxDepth(TreeNode* root) {
    if (!root) return 0;
    return 1 + max(maxDepth(root->left), maxDepth(root->right));
}

int main() {
    //     3
    //    / \
    //   9  20
    //     /  \
    //    15    7
    TreeNode* root = new TreeNode(3);
    root->left  = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left  = new TreeNode(15);
    root->right->right = new TreeNode(7);
    cout << "Max Depth: " << maxDepth(root) << endl; // 3
    return 0;
}
